Some Post Milestone Fixes

robertbartel opened the pull request:
Several fixes made directly to environment during final preparations for milestone test, which now need to be included upstream.  

- Additional details in BMI modules helper build README file.
- Fixes to some relative paths and certain included parameters for several example realization configs.
- Fix of bug in logic for determining from realization config whether a BMI model should use its forcing file directly (issue #254).
- Updating to latest _alt-modular_ checkout.
- Adjust logic in _Bmi_C_Formulation_ to better handle iterating through forcing records, especially near the end of available data, to avoid misleading warnings.

hellkite500 commented on src/realizations/catchment/Bmi_C_Formulation.cpp:
```diff
@@ -74,25 +74,32 @@ inline void Bmi_C_Formulation::get_forcing_data_ts_contributions(time_step_t t_d
+    // Though this is a problem at the moment, because we can't go through forcings in reverse order
+    if (model_epoch_time_s < forcing.get_time_epoch()) {
+        // TODO: think if there is a better way to address this
+        throw std::runtime_error("Cannot get contributions for model time step before current forcing time step");
```
We need to get in the practice of identifying the class/function exceptions are thrown in, and relevant dynamic information, in this case the offending time step `model_epoc_time_s`.

hellkite500 reviewed (commented):
Noting a few things to add based on diffs from the demo repo.
```diff
--- a/src/NGen.cpp
+++ b/src/NGen.cpp
@@ -141,7 +141,7 @@ int main(int argc, char *argv[]) {
                              r_c->get_output_line_for_timestep(output_time_index)+"\n";
         r_c->write_output(output);
         //TODO put this somewhere else.  For now, just trying to ensure we get m^3/s into nexus output
-        response = response * boost::geometry::area(catchment_collection->get_feature(id)->geometry<geojson::multipolygon_t>());
+        response = response * catchment_collection->get_feature(id)->get_property("areasqkm").as_real_number()*1000000;// boost::geometry::area(catchme
         //update the nexus with this flow
         for(auto& nexus : features.destination_nexuses(id))
         {
```

Note, the following is required for building with gcc 4.6.  There are some other potentially cleaner solutions to make this more portable as well.  May need to consider this patch independently.

```diff
--- a/include/utilities/FileStreamHandler.hpp
+++ b/include/utilities/FileStreamHandler.hpp
@@ -10,9 +10,9 @@ namespace utils
       public:
             FileStreamHandler(const char* path) : StreamHandler()
             {
-                auto stream = std::make_shared<std::ofstream>(std::ofstream());
-                stream->open(path, std::ios::trunc);
-                output_stream = stream;
+                output_stream = std::shared_ptr<std::ofstream>( new std::ofstream(path, std::ios::trunc), [](std::ofstream* stream){delete stream;});
+                //output_stream->open(path, std::ios::trunc);
+                //output_stream = stream;
             }
             virtual ~FileStreamHandler(){}
     };
```

These next two are simply additional error reporting that was helpful in debugging, may be worth considering.

```diff
--- a/include/realizations/catchment/Formulation_Manager.hpp
+++ b/include/realizations/catchment/Formulation_Manager.hpp
@@ -209,8 +217,14 @@ namespace realization {
                 utils::StreamHandler output_stream
             ) {
                 auto params = formulation.get_child("params");
-                std::string formulation_type_key =  get_formulation_key(formulation);
-
+                std::string formulation_type_key;
+            try{
+               formulation_type_key =  get_formulation_key(formulation);
+                }
+              catch(std::exception& e)
+              {
+                throw std::runtime_error("Catchment "+identifier+" failed initialization: "+e.what());
+              }
                 boost::property_tree::ptree formulation_config = formulation.get_child("params");

                 auto possible_forcing = tree.get_child_optional("forcing");
```

```diff
--- a/include/realizations/catchment/Formulation_Constructors.hpp
+++ b/include/realizations/catchment/Formulation_Constructors.hpp
@@ -66,7 +66,7 @@ namespace realization {
           return *key;
         }

-        throw std::runtime_error("No valid formulation was described in the passed in tree.");
+        throw std::runtime_error("No valid formulation for "+*key+" was described in the passed in tree.");
     }
 }
```

This patch is likely to be needed eventually if we continue dealing with text input. 
```diff
--- a/include/forcing/Forcing.h
+++ b/include/forcing/Forcing.h
@@ -9,6 +9,7 @@
 #include <fstream>
 #include <iostream>
 #include <boost/algorithm/string.hpp>
+#include <boost/lexical_cast.hpp>
 #include "CSV_Reader.h"
 #include <ctime>
 #include <time.h>
@@ -493,7 +494,7 @@ class Forcing
                     string precip_str = vec[5];

                     //Convert from string to double and from mm/hr to m/s
-                    double precip = atof(precip_str.c_str()) / (1000 * 3600);
+                    double precip = boost::lexical_cast<double>(precip_str) / (1000 * 3600);// atof(precip_str.c_str()) / (1000 * 3600);

                     //Add precip to vector
                     precipitation_rate_meters_per_second_vector.push_back(precip);
@@ -568,20 +569,20 @@ class Forcing
                     AORC_data AORC;

                     //Convert from strings to doubles and add to AORC struct
-                    AORC.APCP_surface_kg_per_meters_squared = atof(APCP_surface_str.c_str());
-                    AORC.DLWRF_surface_W_per_meters_squared = atof(DLWRF_surface_str.c_str());
-                    AORC.DSWRF_surface_W_per_meters_squared = atof(DSWRF_surface_str.c_str());
-                    AORC.PRES_surface_Pa = atof(PRES_surface_str.c_str());
-                    AORC.SPFH_2maboveground_kg_per_kg = atof(SPFH_2maboveground_str.c_str());
-                    AORC.TMP_2maboveground_K = atof(TMP_2maboveground_str.c_str());
-                    AORC.UGRD_10maboveground_meters_per_second = atof(UGRD_10maboveground_str.c_str());
-                    AORC.VGRD_10maboveground_meters_per_second = atof(VGRD_10maboveground_str.c_str());
+                    AORC.APCP_surface_kg_per_meters_squared = boost::lexical_cast<double>(APCP_surface_str);//atof(APCP_surface_str.c_str());
+                    AORC.DLWRF_surface_W_per_meters_squared = boost::lexical_cast<double>(DLWRF_surface_str);//(atof(DLWRF_surface_str.c_str());
+                    AORC.DSWRF_surface_W_per_meters_squared = boost::lexical_cast<double>(DSWRF_surface_str);//atof(DSWRF_surface_str.c_str());
+                    AORC.PRES_surface_Pa = boost::lexical_cast<double>(PRES_surface_str);//atof(PRES_surface_str.c_str());
+                    AORC.SPFH_2maboveground_kg_per_kg = boost::lexical_cast<double>(SPFH_2maboveground_str);//atof(SPFH_2maboveground_str.c_str());
+                    AORC.TMP_2maboveground_K = boost::lexical_cast<double>(TMP_2maboveground_str);//atof(TMP_2maboveground_str.c_str());
+                    AORC.UGRD_10maboveground_meters_per_second = boost::lexical_cast<double>(UGRD_10maboveground_str);//atof(UGRD_10maboveground_str.c_
+                    AORC.VGRD_10maboveground_meters_per_second = boost::lexical_cast<double>(VGRD_10maboveground_str);//atof(VGRD_10maboveground_str.c_

                     //Add AORC struct to AORC vector
                     AORC_vector.push_back(AORC);

                     //Convert precip_rate from string to double
-                    double precip_rate = atof(precip_rate_str.c_str());
+                    double precip_rate = boost::lexical_cast<double>(precip_rate_str);//atof(precip_rate_str.c_str());

                     //Add data to vectors
                     precipitation_rate_meters_per_second_vector.push_back(precip_rate);
```

robertbartel commented:
@hellkite500, I pushed out some changes.  I adjusted the syntax of some the diffs in a few places (e.g., _NGen.cpp_) and removed a few comments.   I also went ahead and applied the change that Donald had made to _ FileStreamHandler.hpp_.  It's a different way to go about it, but I went ahead and addressed it here.

hellkite500 commented:
There is one unexpected test failure (Build and Run NGEN on example data / Build_Run_Model).  I'm ok merging and fixing that in a subsequent issue/PR, or doing it here, either way is fine.

robertbartel commented:
I happen to coincidentally see this was approved right as it was, so I went ahead and merged things before you had a chance to add that comment.  

There are several things that need to be addressed in the tests.  At minimum, the current BMI models need to be move to a public repo, though it also might make sense to remove the testing dependency on those and develop an internal, test-only BMI implementation example that has identity-like functions.  But either way, the last thing you mention can be addressed separately when we get to those tasks.
